### PR TITLE
Loan Management for the Members (Non Profit Module)

### PR DESCRIPTION
**Pull request by shreyashah115**
Summary:

- Changes as per mentioned in #12822.
- Renamed Employee Loan Application to Loan Application and Employee Loan to Loan.
- Added field 'Applicant' and 'Applicant Type' for Employees and Members.
- 'Make Repayment Entry' button visible when Disbursement has been made.
- For repayment, instalment which falls in the current month will be fetched in the Journal Entry for payments.
- Added a field in Loan DocType to display the amount which has been repaid.
- Few fixes in the Loan DocType.
- Added a patch for renaming.

screen shot 2018-03-03 at 12 42 14 am

loan

### CODE REVIEW

**Comment by olamide2** [comment]
Looks Cool

What about instances where members keep money with the NGO as invenstements and expect to get interest on such invenstements as well as pay back at a time in future?

Is that an itch you want to scratch?

**Comment by shreyashah115** [comment]
@olamide2 That needs some discussion, why don't you open a issue for it?

**Comment by kothagunda** [comment]
Hello @umairsy , when will this feature get into Develop. We would like to test it out.

**Comment by umairsy** [comment]
It has been merged already in the develop branch. In the master, you can expect with the release of ERPNext version 11, ETA two weeks (tentative).

**Comment by kothagunda** [comment]
Cool. Thanks @umairsy .

**Comment by kothagunda** [comment]
I installed the developer branch today. But still see Employee loan, instead of Loan docType. Am i missing something. 

image


Installed Apps
ERPNext: v11.x.x-develop (061d912) (develop)
Frappe Framework: v11.x.x-develop (20052f0) (develop)

**Comment by shreyashah115** [comment]
@kothagunda Sorry about the confusion. Yes, it hasn't been merged into develop, yet.

File: erpnext/patches/v11_0/rename_employee_loan_to_loan.py [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+		frappe.rename_doc("DocType", "Employee Loan", "Loan", force=True)
+
+	frappe.reload_doc("hr", "doctype", "loan_application")
+	frappe.reload_doc("hr", "doctype", "loan")
+	frappe.reload_doc("hr", "doctype", "salary_slip_loan")
+
+	for doctype in ['Loan', 'Salary Slip Loan']:
+		if frappe.db.has_column(doctype, 'employee_loan_account'):
+			rename_field(doctype, "employee_loan_account", "loan_account")
+
```
Comment (nabinhait) [inline erpnext/patches/v11_0/rename_employee_loan_to_loan.py]: Patch is required for applicant, applicant_type, loan_application fields as well.

File: erpnext/hr/doctype/loan/loan.py [hunk 2]
```diff
@@ -20,7 +20,10 @@ def validate(self):
 			self.rate_of_interest = frappe.db.get_value("Loan Type", self.loan_type, "rate_of_interest")
 		if self.repayment_method == "Repay Over Number of Periods":
 			self.monthly_repayment_amount = get_monthly_repayment_amount(self.repayment_method, self.loan_amount, self.rate_of_interest, self.repayment_periods)
-
+		if not self.disbursement_date and self.status == "Fully Disbursed":
+			self.disbursement_date = nowdate()
```
Comment (nabinhait) [inline erpnext/hr/doctype/loan/loan.py]: Why nowdate? Ask user to enter.

File: erpnext/hr/doctype/loan/loan.js [hunk 3]
```diff
@@ -46,31 +40,75 @@ frappe.ui.form.on('Employee Loan', {
...
 				frm.trigger("make_jv");
 			})
 		}
+		if (frm.doc.docstatus == 1 && frm.doc.disbursement_date && (frm.doc.applicant_type == 'Member' || frm.doc.repay_from_salary == 0)) {
+			frm.add_custom_button(__('Make Repayment Entry'), function() {
+				frm.trigger("make_repayment_entry");
+			})
+		}
+		if (frm.doc.disbursement_date && !frm.doc.repayment_schedule[0].payment_date) {
+			frm.doc.repayment_schedule[0].payment_date = frm.doc.disbursement_date;
```
Comment (nabinhait) [inline erpnext/hr/doctype/loan/loan.js]: why so?

**Review by nabinhait — CHANGES_REQUESTED** [review]
- Add Loan and Loan Application links in Non-Profit module page

**Review by nabinhait — APPROVED** [review]

File: erpnext/patches/v11_0/rename_employee_loan_to_loan.py (hunk 1 above)
Comment (nabinhait) [reply erpnext/patches/v11_0/rename_employee_loan_to_loan.py]: Fixed via https://github.com/frappe/frappe/commit/3bb36344a6021d8edbe587bbf39c4a22f2d63dc7

**Pull request merged**